### ISSUE TITLE
UGER job array submission

### ISSUE THREAD

**Issue by kshakir** [broadinstitute/cromwell#1491]
Submitting jobs to UGER should use an array:

> You are limited to 100 job in the cluster at once.  This is to encourage the use of Task Arrays.  Task arrays use the variable $SGE_TASK_ID which can be called inside the job script."qsub -t 1-1000 /path/to/jobScript"

via: https://intranet.broadinstitute.org/bits/service-catalog/research-support/job-scheduling

**Comment by katevoss** [comment]
@kshakir I'm unfamiliar with UGER, can you tell me more about what the issue is? Do we have a lot of users with UGER, external and internal?

**Comment by kshakir** [comment]
> Do we have a lot of users with UGER, external and internal?

I can only speak for internal: More and more "Broad" users are using Grid Engine, especially [UGER](https://intranet.broadinstitute.org/bits/service-catalog/compute-resources/high-performance-computing-uger), but not with cromwell, yet.

DSP-methods have been big users of Grid Engine over the past few years, but mostly the older Sun Grid Engine, and with GATK-Queue. Meanwhile BITS is trying to get them to move over to a Univa Grid Engine installation BITS has called UGER.

Unlike the older unlimited setup, UGER's setup has hard limits on the number of concurrent jobs that can be tracked by the Grid Engine scheduler, previously 100, now 1000. With the relaxing of the limit, plus the cromwell's `concurrent-job-limit` feature, this ticket a lower priority imho.

Still, it "would be nice" if just like we submit in batches to JES, we also submitted in batches to other systems that support it including GridEngine/SLURM/PBS/etc.

**Comment by xiaolicbs** [comment]
Hi, Just to follow up on this. I am from the Broad as well. The whole IDMP group at the broad still mainly uses UGER/UGUS Grid Engine for production duties. Many of the tasks will involve dispatching similar jobs thousands of times to the Grid Engine, with will be at least two magnitude faster through a 'task array' mechanism of UGER. I saw support of UGER on cromwell webpage, I wonder if 'task array' is supported by cromwell now?

**Comment by kshakir** [comment]
No, not yet.

A bunch of batching code was put into `JesPollingActor`, but that logic is not available in a generic way. If anyone with cycles does come across this ticket, that might be a good first place to start.

The logic would need to be slightly different. ~JES~ PAPI allows batching _any_ jobs together, while a batch of GridEngine jobs have to use the same resource requirements, aka runtime attributes. Still partial-grouping-for-similar-jobs would still be an improvement over the current single-submit-per-job.

**Comment by tAndreani** [comment]
Hello, does anyone tried to implement a Job Array with LSF scheduler within Cromwell?

I am adapting my pipeline (that uses job array and the LSF scheduler) written in bash for Cromwell. To be sure that every command runs smoothly I have used only 1 sample and everything seems to work fine.

Now I would like to create a job array with multiple samples, e.g. given a file with a list of unique ID (as I usually do in bash with LSF scheduler).

Can anyone share his/her own experience?

Thanks in advance!

**Comment by geoffjentry** [comment]
@tAndreani At least at the moment Cromwell doesn't play well with array jobs, as the individual jobs are parallelized before being sent to the backend, which is where the concept of an array job would be understood.

**Comment by tAndreani** [comment]
@geoffjentry thanks for the quick reply! I am using now scatter and it works pretty well, meaning that multiple samples are running in parallel for the same task.. My question is: how many samples the scatter can run simultaneously? As a test am doing with 2 samples and they both run at the same time.. but am wondering if there is a limit (e.g. 100 or 200 etc..). This will help me to understand the time of the run of my entire workflow.

Thanks in advance for any reply

**Comment by geoffjentry** [comment]
@tAndreani in terms of cromwell there's not a limit per se, although your underlying backend might get grouchy at you if you wind up submitting too many jobs. You could set the `concurrent-job-limit` config field to help with this if you do wind up with issues

**Comment by tAndreani** [comment]
A quick question: if I run in local my workflow and I parallelize with several sample using scatter, Cromwell takes all the CPU that we have in our machine. I then used this otion in every task:

```
 runtime {
  docker_user: "ngs"
  cpu: 6
  }
```

But this does not work, and all the CPU are taken at the same time. And i have this error:

`[warn] BackgroundConfigAsyncJobExecutionActor [7e5755bcscMeth.mapping:0:1]: Unrecognized runtime attribute keys: cpu`
Is there any explanation for this?

Thanks in advance.

**Comment by geoffjentry** [comment]
@tAndreani `cpu` isn't a supported `runtime` value in the local backend (really, very few of them are). With the local backend Cromwell won't limit the resource usage of any job. We view the local backend more as a developmental/experimentation tool than something to be used in a real world scenario, so there's less attention paid to that sort of thing. The concurrent job limit could be used here as well.
